File: dsa-problems/TreesDP/solutions/javascript/TDP-002-tree-diameter-dp.js
const readline = require("readline");

class Solution {
  constructor() {
    this.tree = [];
    this.diameter = 0;
  }

  treeDiameter(n, edges) {
    this.tree = Array.from({ length: n + 1 }, () => []);
    this.diameter = 0;

    for (const [u, v] of edges) {
      this.tree[u].push(v);
      this.tree[v].push(u);
    }

    this.dfs(1, -1);
    return this.diameter;
  }

  dfs(node, parent) {
    let max1 = 0,
      max2 = 0;

    for (const child of this.tree[node]) {
      if (child === parent) continue;

      const childDepth = this.dfs(child, node);

      if (childDepth > max1) {
        max2 = max1;
        max1 = childDepth;
      } else if (childDepth > max2) {
        max2 = childDepth;
      }
    }

    this.diameter = Math.max(this.diameter, max1 + max2);
    return max1 + 1;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const lines = [];
rl.on("line", (line) => {
  lines.push(line);
});

rl.on("close", () => {
  const n = parseInt(lines[0]);

  const edges = [];
  for (let i = 1; i < n; i++) {
    const [u, v] = lines[i].split(" ").map(Number);
    edges.push([u, v]);
  }

  const solution = new Solution();
  const result = solution.treeDiameter(n, edges);

  console.log(result);
});

File: dsa-problems/TreesDP/solutions/java/TDP-002-tree-diameter-dp.java
import java.util.*;

class Solution {
    private List<Integer>[] tree;
    private int diameter = 0;

    public int treeDiameter(int n, int[][] edges) {
        tree = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            tree[i] = new ArrayList<>();
        }

        for (int[] edge : edges) {
            tree[edge[0]].add(edge[1]);
            tree[edge[1]].add(edge[0]);
        }

        dfs(1, -1);
        return diameter;
    }

    private int dfs(int node, int parent) {
        int max1 = 0, max2 = 0;

        for (int child : tree[node]) {
            if (child == parent) continue;

            int childDepth = dfs(child, node);

            if (childDepth > max1) {
                max2 = max1;
                max1 = childDepth;
            } else if (childDepth > max2) {
                max2 = childDepth;
            }
        }

        diameter = Math.max(diameter, max1 + max2);
        return max1 + 1;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        int[][] edges = new int[n - 1][2];
        for (int i = 0; i < n - 1; i++) {
            edges[i][0] = sc.nextInt();
            edges[i][1] = sc.nextInt();
        }

        Solution solution = new Solution();
        int result = solution.treeDiameter(n, edges);

        System.out.println(result);
        sc.close();
    }
}
